### ISSUE TITLE
Why some regions have the same signal value in *12KB_STITCHED_TSS_DISTAL_ENHANCER_REGION_MAP.txt on my data?

### ISSUE THREAD

**Issue by Tang-RH** [stjude/ROSE#23]
Dear ROSE developers,
    I ran ROSE 1.3.1 on my data recently. The command is : 

`ROSE_main.py  -g HG38  -i SRR17567976_rmchrM_sorted_rmdup_peaks_bl_vtss.gff -r SRR17567976_rmchrM_sorted_rmdup.bam -o SRR17567979_rmchrM_sorted_rmdup_peaks_bl_vtss  -s 12500 -t 2500 `

However the Plot_point.png in the result folder is broken line, and I found that some regions had the same signal value in 12KB_STITCHED_TSS_DISTAL_ENHANCER_REGION_MAP.txt. The first 12 rows are showed below:

REGION_ID	CHROM	START	STOP	NUM_LOCI	CONSTITUENT_SIZE	SRR17567976_rmchrM_sorted_rmdup.bam
36_._lociStitched	chr11	47982919	48081163	36	14550	53012.4624
20_._lociStitched	chr6	16659618	16728498	20	9913	34150.704
11_._lociStitched	chr5	95020044	95088151	11	4421	16561.8112
13_._lociStitched	chr8	27327938	27390341	13	5155	20561.788500000002
13_._lociStitched	chr17	75295218	75352547	13	3916	20561.788500000002
10_._lociStitched	chr6	7112498	7167165	10	5050	12622.8816
19_._lociStitched	chr15	78034284	78088444	19	8883	39324.74
20_._lociStitched	chr5	78496236	78549387	20	14829	34150.704
11_._lociStitched	chr11	117959641	118012359	11	5031	16561.8112
17_._lociStitched	chr4	158772306	158821814	17	8416	32536.6576
13_._lociStitched	chr4	89283495	89332242	13	5265	20561.788500000002

the .png is as follow:
![SRR17567976_rmchrM_sorted_rmdup_peaks_bl_vtss_Plot_points](https://user-images.githubusercontent.com/89056018/208926991-6ad831ed-ce34-4ef1-8d04-e8610767f9dc.png)

Then I ran the previous version of ROSE on the same .bam and .bed files, the result seemed normal. The .png is as follow:
![SRR17567976_rmchrM_sorted_rmdup_peaks_bl_vtss_Plot_points](https://user-images.githubusercontent.com/89056018/208928462-33fb84f2-6514-453a-b087-b3823b08290c.png)

Did I make mistakes when ran the ROSE 1.3.1? Why I got the strange result?

Thanks and best regards

**Comment by AndrewC160** [comment]
This comment is pretty old but I was having this same issue and I think I have found the problem after delving into the source code a bit.

The TL;DR version is that you need to include unique ID's for each enhancer region: if you look at @Tang-RH 's IDs they are not unique (36_.lociStitched shows up twice, and has the same scores for each BAM file). I think this problem ultimately stems from the README.md, which states that unique enhancer IDs must be present in column 2, but in my data it actually appear to be pulled from column 3 (in my case I filled in the column with "Enhancer" for each entry in col. 3, so no unique values). I am also open to the possiblity my GFF formatting was wrong, but I'll check that shortly

ROSE's enhancer IDs are formatted as <number_of_stitched_regions>_<unique_id>_lociStitched. If the input IDs are not unique (in @Tang-RH 's case it looks like they were empty),  ID's become <number_of_stitched_regions>_lociStitched. When scores are placed in a dictionary during mapBamToGFF(), the score of the first stitched peak with with 36 regions will be used and all subsequent stitched peaks will get the same score (since they're all called 36_.lociStitched).

This doesn't crash things and still gives reasonable-looking results like in plot 1 (above) because the most pronounced enhancers problably have an abnormally large number of stitched regions, so they may still have unique IDs. When you get to the less pronounced enhancers with 3 or 4 peaks, though, you start seeing duplicates which are plotted in the rank plots as horizontal lines.

**Comment by AndrewC160** [comment]
Hi @dorienpastoors ,

I haven't poked around in this code for a while but I do somewhat remember what we did, do you have a few example lines for your input tables as well as the input file names? My guess is you're right that there's another dictionary issue and there's either a filename or set of column values that get turned into an identifier which is not unique in your case.

**Comment by dorienpastoors** [comment]
Hi @AndrewC160,

thanks for your fast reply. 
My input GFF file looks as follows:

| 1    | 2      | 3 | 4     | 5     | 6       | 7 | 8 | 9      |
| ---- | ------ | - | ----- | ----- | ------- | - | - | ------ |
| chr1 | peak_1 | 1 | 10089 | 10333 | 2.9251  | . | . | peak_1 |
| chr1 | peak_2 | 2 | 11103 | 11454 | 3.76889 | . | . | peak_2 |
| chr1 | peak_3 | 3 | 28802 | 29415 | 6.55833 | . | . | peak_3 |
| chr1 | peak_4 | 4 | 29664 | 29988 | 3.37501 | . | . | peak_4 |
| chr1 | peak_5 | 5 | 30573 | 30846 | 2.85967 | . | . | peak_5 |
| chr1 | peak_6 | 6 | 32245 | 33018 | 28.2512 | . | . | peak_6 |
| chr1 | peak_7 | 7 | 34302 | 34995 | 8.18636 | . | . | peak_7 |

The file does not contain a header, ive numbered the columns just for this markdown table, so the .gff file starts like this:

chr1	peak_1	1	10089	10333	2.9251	.	.	peak_1
chr1	peak_2	2	11103	11454	3.76889	.	.	peak_2
chr1	peak_3	3	28802	29415	6.55833	.	.	peak_3

the columns in this file are pretty self explanatory, but they are : c("seqnames", "regionID","feature", "start", "end", "fold_enrichment","strand", "frame", "regionID"). I made it from a macs2 file. Feature is just a column I made from the regionID that does not contain the prefix peak_ , otherwise its the same ID. (I tried some different things already, including the feature column being the same as the regionID column but this does not make a difference). 

The command I use **with** the control is:
```
ROSE_main.py -g hg19 -i path_to_gff.gff \
 -r path_to_p300_chip.bam \
 -o output_dir \
 -c path_to_input.bam
```
and the command I use **without** control is:

```
ROSE_main.py -g hg19 -i path_to_gff.gff \
 -r path_to_p300_chip.bam \
 -o output_dir 
```
The last issue in my previous comment makes me think it might not be the dictionary per se - the weird thing is that I do get unique scores the chip.bam for the command **without** control, and I get unique scores for the input.bam in the first command. So maybe it has to do with how these functions are called and that in case of an input file is given, there is some if statement somewhere that runs the dictionary function with different set of keys (the # of constituent peaks instead of the peakID) in case of having an input file in the input command specifically for the chip.bam file. 

PS you mean the input GFF file right? not the headers of bam files or something.

**Comment by AndrewC160** [comment]
Hello @dorienpastoors ,

Sorry I responded quickly at first then fell off the face of the earth for a while!

Not having used ROSE in a few months I don't know off the bat why your data would have an issue, but here is a command line call and the head of my input GFF file that has worked in a pipeline I wrote for our data. You could try replicating these formats with your own data and see if that helps. A few notes that _may_ be important (no idea but couldn't hurt to try):

1) my input GFF was 8 columns
2) my enhancer IDs were "Enhancer_1" rather than 1; shot in the dark but maybe your column 3 gets turned into an integer value and that changes things? Again, just a guess.
3) My score column was defaulted to a value of 1, so if your scores aren't important maybe change that.
4) I only have one "empty" column in 7.

Hope this helps!

Command:

ROSE_main.py \\
-i /N/p/asclab/ASC-cutNrun/data/pools/OS525_K27Ac_NONE/superenhancers/OS525_K27Ac_NONE_peaks.gff \\
-o /N/p/asclab/ASC-cutNrun/data/pools/OS525_K27Ac_NONE/superenhancers \\
-g HG38 \\
-r /N/p/asclab/ASC-cutNrun/data/samples/OS525_K27Ac_NONE_1/align/OS525_K27Ac_NONE_1_hg38_sub120bp.bam \\
-b /N/p/asclab/ASC-cutNrun/data/samples/OS525_K27Ac_NONE_2/align/OS525_K27Ac_NONE_2_hg38_sub120bp.bam \\
-c /N/p/asclab/ASC-cutNrun/data/samples/OS525_IgG_NONE_1/align/OS525_IgG_NONE_1_hg38_sub120bp.bam \\
-s 12500 \\
-t 10000

Input table (no header, column numbers are just here for ease of reading)

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ACLUGS~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ACLUGS~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


1 | 2 | 3 | 4 | 5 | 6 | 7 | 8
-- | -- | -- | -- | -- | -- | -- | --
chr1 | OS525_K27Ac_NONE_peak_1 | Enhancer_1 | 28970 | 29262 | 1 | . | OS525_K27Ac_NONE_peak_1
chr1 | OS525_K27Ac_NONE_peak_2 | Enhancer_2 | 199425 | 200346 | 1 | . | OS525_K27Ac_NONE_peak_2
chr1 | OS525_K27Ac_NONE_peak_3 | Enhancer_3 | 778404 | 778601 | 1 | . | OS525_K27Ac_NONE_peak_3
chr1 | OS525_K27Ac_NONE_peak_4 | Enhancer_4 | 826614 | 826808 | 1 | . | OS525_K27Ac_NONE_peak_4
chr1 | OS525_K27Ac_NONE_peak_5 | Enhancer_5 | 827117 | 827385 | 1 | . | OS525_K27Ac_NONE_peak_5
chr1 | OS525_K27Ac_NONE_peak_6 | Enhancer_6 | 940513 | 940716 | 1 | . | OS525_K27Ac_NONE_peak_6
chr1 | OS525_K27Ac_NONE_peak_7 | Enhancer_7 | 1013024 | 1013946 | 1 | . | OS525_K27Ac_NONE_peak_7
chr1 | OS525_K27Ac_NONE_peak_8 | Enhancer_8 | 1157738 | 1158475 | 1 | . | OS525_K27Ac_NONE_peak_8
chr1 | OS525_K27Ac_NONE_peak_9 | Enhancer_9 | 1330932 | 1331181 | 1 | . | OS525_K27Ac_NONE_peak_9



</body>

</html>

**Comment by dorienpastoors** [comment]
Hi andrew, 

No worries - thank you so much for taking the time to reply. Yes - before I had peak_1, peak_2, ...,  peak_n in the third column instead of just an integer but this did not make a difference to the output. 
I can replicate your GFF file indeed! thanks for sharing! A collegue of mine also suggested to use a BED file as input instead of GFF to maybe circumvent this issue as well.
The workaround I have now is to just run all additional files in the -bams argument (my goal is to compare signal in different treatments with each other, so i have like 15 files per call) and I'm also including the input file there. So i'm not running the input file in the -c flag but with the -bams flag. I then just calculate the normalised ranks on signal - input in R. This works well and I can see the output makes sense now :)
I mean that the output looks good now - running a lot of bam files in the -bams argument will occupy all available cores as there is no parallelization control in the current script. - just if anyone is reading this who has a similar issue.
